Change detector integrate

kegluneq opened the pull request.

Commit:
feat(dart/transform): Use the best available Change Detectors

Enable pregenerated (for Dart) and JIT (for Js) change detectors when
possible. Previously we would always use `DynamicChangeDetector`s, but
these cause megamorphic calls and are therefore much slower.

Closes #502

vsavkin commented on modules/angular2/src/change_detection/change_detection.ts:
```diff
@@ -68,17 +68,20 @@ export var defaultPipes = {
 
 export var preGeneratedProtoDetectors = {};
 
+export const PROTO_CHANGE_DETECTOR_KEY = 'ProtoChangeDetectors';
```
You should use something like OpaqueToken here.

vsavkin commented on modules/angular2/src/core/application.ts:
```diff
@@ -118,7 +118,7 @@ function _injectorBindings(appComponentType): List<Type | Binding | List<any>> {
     TemplateResolver,
     bind(PipeRegistry).toValue(defaultPipeRegistry),
-    bind(ChangeDetection).toClass(DynamicChangeDetection),
+    bind(ChangeDetection).toClass(PreGeneratedChangeDetection),
```
To enable this we need to do the following two things:

* Ports all tests to make sure that the new detector actually works the same way.
* Enable JIT detection for javascript. We should do that by conditionally setting the detector: JIT for JS, PreGenerated for Dart.

vsavkin replied:
To enable this we need to do the following two things:

* Ports all tests to make sure that the new detector actually works the same way.
* Enable JIT detection for javascript. We should do that by conditionally setting the detector: JIT for JS, PreGenerated for Dart.

naomiblack commented:
Also looks like tests are failing on clang-format. Make sure you've updated to 1.0.15

vsavkin commented:
@kegluneq could you update this PR?

kegluneq commented:
Yep, this is my next focus. I just pushed an update incorporating all the recent changes to pre-generated change detectors -- now I want to work on properly enabling the pregen vs. JIT change detectors depending on when they're available.

Commit:
fix(change detect): Fix bug in JIT change detectors

Fix the "ifChangedGuard" code in the JIT change detector which was
incorrectly refactored in a2770c8a5259aacc64f21170a93b5124f55511b8.

Commit:
fix(dynamic_component_loader): Fix for ts2dart issue

Fix a build break by declaring a type for `locals`.
See https://github.com/angular/ts2dart/wiki/Builtin-fa%C3%A7ade-methods#fixing-untyped-property-access-errors
for some context.
